packages/dapi/lib/grpcServer/handlers/blockheaders-stream/subscribeToBlockHeadersWithChainLocksHandlerFactory.js

```diff
@@ -0,0 +1,193 @@
+ * @param {ZmqClient} zmqClient
+ * @param {subscribeToNewBlockHeaders} subscribeToNewBlockHeaders
+ * @return {subscribeToBlockHeadersWithChainLocksHandler}
+ */
+function subscribeToBlockHeadersWithChainLocksHandlerFactory(
+  getHistoricalBlockHeadersIterator,
+  coreAPI,
+  zmqClient,
+  subscribeToNewBlockHeaders,
+) {
+  /**
+   * @typedef subscribeToBlockHeadersWithChainLocksHandler
+   * @param {grpc.ServerWriteableStream<BlockHeadersWithChainLocksRequest>} call
+   */
+  async function subscribeToBlockHeadersWithChainLocksHandler(call) {
+    const { request } = call;
+
+    let fromBlockHash = Buffer.from(request.getFromBlockHash()).toString('hex');
+    const fromBlockHeight = request.getFromBlockHeight();
+    const count = request.getCount();
+
+    const newHeadersRequested = count === 0;
+
+    const acknowledgingCall = new AcknowledgingWritable(call);
+
+    const mediator = new ProcessMediator();
+
+    mediator.on(
+      ProcessMediator.EVENTS.BLOCK_HEADERS,
+      async (blockHeaders) => {
+        await sendBlockHeadersResponse(acknowledgingCall, blockHeaders);
+      },
+    );
+
+    mediator.on(
+      ProcessMediator.EVENTS.CHAIN_LOCK_SIGNATURE,
+      async (clSig) => {
+        await sendClSigResponse(acknowledgingCall, clSig);
+      },
+    );
+
+    if (newHeadersRequested) {
+      subscribeToNewBlockHeaders(mediator, zmqClient, coreAPI);
+    }
+
+    // If block height is specified instead of block hash, we obtain block hash by block height
+    if (fromBlockHash === '') {
+      if (fromBlockHeight === 0) {
+        throw new InvalidArgumentGrpcError('minimum value for `fromBlockHeight` is 1');
+      }
+
+      // we don't need to check bestBlockHeight because getBlockHash throws
+      // an error in case of wrong height
+      try {
+        fromBlockHash = await coreAPI.getBlockHash(fromBlockHeight);
+      } catch (e) {
+        if (e.code === -8) {
+          // Block height out of range
+          throw new NotFoundGrpcError('fromBlockHeight is bigger than block count');
+        }
+
+        throw e;
+      }
+    }
+
+    let fromBlock;
+
+    try {
+      fromBlock = await coreAPI.getBlock(fromBlockHash);
+    } catch (e) {
+      // Block not found
+      if (e.code === -5) {
+        throw new NotFoundGrpcError('fromBlockHash is not found');
+      }
+
+      throw e;
+    }
+
+    if (fromBlock.confirmations === -1) {
+      throw new NotFoundGrpcError(`block ${fromBlockHash} is not part of the best block chain`);
+    }
+
+    const bestBlockHeight = await coreAPI.getBestBlockHeight();
+
+    const historicalCount = count === 0 ? bestBlockHeight - fromBlock.height + 1 : count;
+
+    if (fromBlock.height + historicalCount > bestBlockHeight + 1) {
+      throw new InvalidArgumentGrpcError('`count` value exceeds the chain tip');
+    }
+
+    const bestChainLock = await coreAPI.getBestChainLock();
+    await sendClSigResponse(acknowledgingCall, Buffer.from(bestChainLock.signature, 'hex'));
```

Comment:
Core might not have chainlocks yet (for example is just started and didn't receive it from peers) so you need to handle it.

Reply:
We need to pass the whole chainlock object to the client, not just the signature. Otherwise client won't be able to verify it.

Reply:
We probably need to use latest received chainlock. so when you start tx-filter-stream process you get it with RPC and then listen. So here you can use the latest one.

Reply:
>We need to pass the whole chainlock object to the client, not just the signature. Otherwise client won't be able to verify it.

Oh, I referred to the ticket which says that endpoint should return CLSIG. Currently, protobuffer messages also suggesting usage of signatures. 
Hence, I suggest to handle it in the next iteration together with the protobuf specs, when it comes to actual verification. What do you think?

Reply:
>We probably need to use latest received chainlock. so when you start tx-filter-stream process you get it with RPC and then listen. So here you can use the latest one.

I'm looking at the core RPC spec and don't see a method for retrieving the latest received chainlock. Can you please elaborate?

Reply:
CLSIG is a name of Core p2p message which represented as an object with message, signature and so on in Core RPC

Reply:
In [dapi-grpc](https://github.com/dashevo/platform/blob/c7850e34472f443da3a9ef7d18be43555463c20b/packages/dapi/lib/grpcServer/handlers/blockheaders-stream/subscribeToBlockHeadersWithChainLocksHandlerFactory.js#L17) generated protobuffer entity called `ChainLockSignatureMessages`. And also it expects array of signatures, while in fact we are sending only single item in response. Perhaps this is also should be updated according to newest logic?

Reply:
Yeah, I guess we can rename it. It accepts an array of bytes because we send serialized object, like instant locks. It's not obvious though.

Reply:
Fixed